subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/AbstractValidatingProperty.java

```diff
@@ -40,16 +41,18 @@ public AbstractValidatingProperty(String propertyName, PropertyValue value, bool
         this.validationAction = validationAction;
     }
 
-    public static void reportValueNotSet(String propertyName, TypeValidationContext context) {
+    public static void reportValueNotSet(String propertyName, TypeValidationContext context, boolean hasConfigurableValue) {
         context.visitPropertyProblem(problem -> {
             problem.withId(ValidationProblemId.VALUE_NOT_SET)
                 .reportAs(Severity.ERROR)
                 .forProperty(propertyName)
                 .withDescription("doesn't have a configured value")
                 .happensBecause("This property isn't marked as optional and no value has been configured")
-                .addPossibleSolution(() -> "Assign a value to '" + propertyName + "'")
-                .addPossibleSolution(() -> "Mark property '" + propertyName + "' as optional")
                 .documentedAt("validation_problems", "value_not_set");
+            if (hasConfigurableValue) {
+                problem.addPossibleSolution(() -> "Assign a value to '" + propertyName + "'");
+            }
```

Comment:
❓ I wonder if we could say something here when there's no configurable value... Marking the property as `@Optional` is for the author of the task. For a software developer who just uses the build this is not helpful. The ultimate solution is to explain where the value comes from, but we don't currently have provenance tracking as a feature to do that properly. Yet I think we should be able to say something here already, perhaps along the lines of "This is a calculated value, make sure it can calculate to a valid value"?